sklearn/utils/extmath.py

```diff
@@ -255,37 +257,64 @@ def randomized_range_finder(
     analysis
     A. Szlam et al. 2014
     """
+    xp, is_array_api_compliant = get_namespace(A)
     random_state = check_random_state(random_state)
 
     # Generating normal random vectors with shape: (A.shape[1], size)
-    Q = random_state.normal(size=(A.shape[1], size))
-    if hasattr(A, "dtype") and A.dtype.kind == "f":
-        # Ensure f32 is preserved as f32
-        Q = Q.astype(A.dtype, copy=False)
+    # XXX: generate random number directly from xp if it's possible
+    # one day.
+    Q = xp.asarray(random_state.normal(size=(A.shape[1], size)))
+    if hasattr(A, "dtype") and xp.isdtype(A.dtype, kind="real floating"):
+        # Use float32 computation and components if A has a float32 dtype
+        Q = xp.astype(Q, A.dtype, copy=False)
+
+    # Move Q to device if needed only after converting to float32 if needed to
+    # avoid allocating unnecessary memory on the device.
+    Q = to_device(Q, getattr(A, "device", None))
```

Comment:
I think this whole section can be simplified to:

```python
    # XXX: generate random number directly from xp if it's possible
    # one day.
    Q_dtype = None
    if hasattr(A, "dtype") and xp.isdtype(A.dtype, kind="real floating"):
        # Use float32 computation and components if A has a float32 dtype
        Q_dtype = A.dtype

    Q = xp.asarray(
        random_state.normal(size=(A.shape[1], size)), dtype=Q_dtype, device=device(A)
    )
```

where `device` is from `sklearn/utils/_array_api.py`. `device` would need to be updated to:

```python
def device(x):
    if isinstance(x, (numpy.ndarray, numpy.generic)) or issparse(x):
        return "cpu"
    return x.device
```

This also means we do not need a `to_device` function anymore.

Reply:
It's not clear that that's the case, because asarray() in the spec says that it might not support downcasts. See the discussion at https://github.com/data-apis/array-api/issues/647 (I believe this may be discussed at the next arrays consortium meeting).

Reply:
Thanks for your input @asmeurer. I added a note in 4093a6c for now.